File: CS/Form1.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using System.Windows.Forms;
using Northwind;
using DevExpress.XtraGrid.Views.Grid;

namespace FilteredAssociation {
    public partial class Form1 : Form {
        public Form1() {
            InitializeComponent();
        }

        private void gridView1_MasterRowGetChildList(object sender, DevExpress.XtraGrid.Views.Grid.MasterRowGetChildListEventArgs e) {
            GridView masterGrid = sender as GridView;
            Category category = (Category)masterGrid.GetRow(e.RowHandle);
            e.ChildList = category.AvailableProducts;
        }
    }
}

File: VB/Form1.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Text
Imports System.Windows.Forms
Imports Northwind
Imports DevExpress.XtraGrid.Views.Grid

Namespace FilteredAssociation
	Partial Public Class Form1
		Inherits Form
		Public Sub New()
			InitializeComponent()
		End Sub

		Private Sub gridView1_MasterRowGetChildList(ByVal sender As Object, ByVal e As DevExpress.XtraGrid.Views.Grid.MasterRowGetChildListEventArgs) Handles gridView1.MasterRowGetChildList
			Dim masterGrid As GridView = TryCast(sender, GridView)
			Dim category As Category = CType(masterGrid.GetRow(e.RowHandle), Category)
			e.ChildList = category.AvailableProducts
		End Sub
	End Class
End Namespace